test/Microsoft.AspNet.Mvc.FunctionalTests/JsonPatchTest.cs

```diff
@@ -0,0 +1,146 @@
+                Method = new HttpMethod("PATCH"),
+                RequestUri = new Uri(url)
+            };
+
+            // Act
+            var response = await client.SendAsync(request);
+
+            // Assert
+            var body = await response.Content.ReadAsStringAsync();
+            var customer = JsonConvert.DeserializeObject<Customer>(body);
+            Assert.Equal("Name2", customer.Orders[2].OrderName);
+        }
+
+
+        [Theory]
+        [InlineData("http://localhost/JsonPatch/JsonPatchWithModelState")]
+        [InlineData("http://localhost/JsonPatch/JsonPatchWithModelStateAndPrefix?prefix=Patch")]
+        public async Task JsonPatch_MultipleValidOperations_Success(string url)
+        {
+            // Arrange
+            var server = TestHelper.CreateServer(_app, SiteName, _configureServices);
+            var client = server.CreateClient();
+
+            var input = "[{ \"op\": \"add\", \"path\": \"Customer/Orders/2\", " +
+               "\"value\": { \"OrderName\": \"Name2\" }}, {\"op\": \"copy\", \"from\": \"Customer/Orders/2\", " +
+                "\"path\": \"Customer/Orders/3\" }, {\"op\": \"replace\", \"path\": \"Customer/Orders/2/OrderName\", " +
+                "\"value\": \"ReplacedName\" }]";
+            var request = new HttpRequestMessage
+            {
+                Content = new StringContent(input, Encoding.UTF8, "application/json-patch+json"),
+                Method = new HttpMethod("PATCH"),
+                RequestUri = new Uri(url)
+            };
+
+            // Act
+            var response = await client.SendAsync(request);
+
+            // Assert
+            var body = await response.Content.ReadAsStringAsync();
+            var customer = JsonConvert.DeserializeObject<Customer>(body);
+            Assert.Equal("ReplacedName", customer.Orders[2].OrderName);
+            Assert.Equal("Name2", customer.Orders[3].OrderName);
+        }
+
+        public static IEnumerable<object[]> InvalidJsonPatchData
+        {
+            get
+            {
+                return new[]
+                {
+                    new object[] {
+                        "http://localhost/JsonPatch/JsonPatchWithModelStateAndPrefix?prefix=Patch",
+                        "[{ \"op\": \"add\", \"path\": \"Customer/Orders/5\", " +
+                        "\"value\": { \"OrderName\": \"Name5\" }}]",
+                        "{\"Patch.JsonPatchDocument`1\":[\"For operation 'add' on array property at path " +
+                        "'Customer/Orders/5', the index is larger than the array size.\"]}"
+                    },
+                    new object[] {
+                        "http://localhost/JsonPatch/JsonPatchWithModelState",
+                        "[{ \"op\": \"add\", \"path\": \"Customer/Orders/5\", " +
+                        "\"value\": { \"OrderName\": \"Name5\" }}]",
+                        "{\"JsonPatchDocument`1\":[\"For operation 'add' on array property at path " +
```

Comment:
Can we do better than `JsonPatchDocument` for the error? I would think that's we'd use the property path that failed - or at least the inner-type of the document.